extensions/amp-stream-gallery/1.0/stream-gallery.js

```diff
@@ -0,0 +1,217 @@
+/**
+ * Copyright 2019 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as Preact from '../../../src/preact';
+import {BaseCarousel} from '../../amp-base-carousel/1.0/base-carousel';
+import {
+  useCallback,
+  useLayoutEffect,
+  useMemo,
+  useRef,
+  useState,
+} from '../../../src/preact';
+import {useStyles} from './stream-gallery.jss';
+
+const DEFAULT_MEASUREMENT = {
+  visibleCount: 1,
+  maxContainerWidth: Number.MAX_VALUE,
+};
+
+/**
+ * @param {!StreamGalleryDef.Props} props
+ * @return {PreactDef.Renderable}
+ */
+export function StreamGallery({
+  arrowPrev: customArrowPrev,
+  arrowNext: customArrowNext,
+  children,
+  extraSpace,
+  insetArrowVisibility,
+  loop,
+  maxItemWidth = Number.MAX_VALUE,
+  minItemWidth = 1,
+  maxVisibleCount = Number.MAX_VALUE,
+  minVisibleCount = 1,
+  outsetArrows,
+  peek = 0,
+  snap,
+  style,
+  ...rest
+}) {
+  const galleryRef = useRef(null);
+  const [measurements, setMeasurements] = useState(DEFAULT_MEASUREMENT);
+  const arrowPrev = useMemo(
+    () =>
+      customArrowPrev || <DefaultArrow by={-1} outsetArrows={outsetArrows} />,
+    [customArrowPrev, outsetArrows]
+  );
+  const arrowNext = useMemo(
+    () =>
+      customArrowNext || <DefaultArrow by={1} outsetArrows={outsetArrows} />,
+    [customArrowNext, outsetArrows]
+  );
+
+  const measure = useCallback(
+    (containerWidth) =>
+      getVisibleCount(
+        maxItemWidth,
+        minItemWidth,
+        maxVisibleCount,
+        minVisibleCount,
+        children.length,
+        peek,
+        containerWidth
+      ),
+    [
+      maxItemWidth,
+      minItemWidth,
+      maxVisibleCount,
+      minVisibleCount,
+      children.length,
+      peek,
+    ]
+  );
+
+  // Adjust visible slide count when container size or parameters change.
+  useLayoutEffect(() => {
+    const node = galleryRef.current;
+    if (!node) {
+      return;
+    }
+    const observer = new ResizeObserver((entries) => {
```

Comment:
Optional: can we make it such that that the ResizeObserver is only instantiated once, even if the ref or measure object change?

Reply:
I modeled this after the IntersectionObserver usage in Timeago. This was the [discussion thread at the time](https://github.com/ampproject/amphtml/pull/26507#discussion_r384880297) wherein we decided not to share across instances if that is relevant.

It sounds like your suggestion is to share across renders of the same instance, but I'm confused why we would want this to be independent of `measure` which itself depends on many props. Can you clarify?

Reply:
I just measured and in latest Chrome it only took ~0.000001 seconds to instantiate a new ResizeObserver...so feel free to actually ignore this comment entirely :)

> It sounds like your suggestion is to share across renders of the same instance

Yep! 

> Can you clarify?

Its just a hypothetical ideal. The ResizeObserver is still pointed at the same element, so there is no reason to make a new one. The only time it would be required to futz with RO is if galleryRef.current (the node) changed in which case you'd want to unobserve the old elem and observe the new one.

The RO callback always needs to call the latest version of measure, but that doesn't require a whole new RO.